### ISSUE TITLE
zfs pool (harddrive) notifications

### ISSUE THREAD

**Issue by thefiredragon** [netdata/netdata#5590]
Hello, I'm not a good programmer and would like to ask if someone could bring up a feature which alert me when a pool is failed or degraded 'cause a hard drive riches end of life? 

best Regards, Dan

**Comment by cakrit** [comment]
@vlvkobal can you help here?

**Comment by thefiredragon** [comment]
Sorry I can't 'cause my skills are more at php. :(

Chris Akritidis <notifications@github.com> schrieb am Mo., 29. Apr. 2019,
14:48:

**Comment by cakrit** [comment]
Hi @thefiredragon , don't worry about writing the code, but we do need you to tell us how we'd detect that error: i.e. how would you check it from the command line or what file you would read for a specific string etc.

**Comment by thefiredragon** [comment]
With zpool status , the state tell us if the pool is okay or when there's a
failure,, this looks like this:


root@DansHomeServer:~# zpool status
  pool: vstorage
 state: ONLINE
  scan: scrub repaired 408K in 2h39m with 0 errors on Sun Apr 14 03:03:58
2019
config:

NAME                                           STATE     READ WRITE CKSUM
vstorage                                       ONLINE       0     0     0
  raidz1-0                                     ONLINE       0     0     0
    ata-ST2000LX001-1RG174_WDZAG9YT            ONLINE       0     0     0
    ata-ST2000LX001-1RG174_WDZBB0WQ            ONLINE       0     0     0
    ata-ST2000LX001-1RG174_WDZBG6R1            ONLINE       0     0     0
    ata-ST2000LX001-1RG174_WDZFDAQZ            ONLINE       0     0     0
    ata-ST2000LX001-1RG174_WDZFG9XB            ONLINE       0     0     0
logs
  ata-SSDSC2BB120G7R_PHDV746001VU150MGN-part2  ONLINE       0     0     0
cache
  ata-SSDSC2BB120G7R_PHDV746001VU150MGN-part1  ONLINE       0     0     0

errors: No known data errors

Am Fr., 3. Mai 2019 um 09:43 Uhr schrieb Chris Akritidis <
notifications@github.com>:

**Comment by vlvkobal** [comment]
Could you point it out how it looks when there are problems with disks?

**Comment by thefiredragon** [comment]
Good explain you could find Here:

https://docs.oracle.com/cd/E19253-01/819-5461/gbcve/index.html


At the moment all Pools okay so i cant send examples ..

Vladimir Kobal <notifications@github.com> schrieb am Fr., 3. Mai 2019,
11:05:

**Comment by vlvkobal** [comment]
We need to investigate [ZFS library](https://github.com/zfsonlinux/zfs) and find a convenient way to get this information. We are putting this into our backlog.

**Comment by mfundul** [comment]
There is no ZFS library for linux yet though.

**Comment by thefiredragon** [comment]
There’s a big port for zfs called zfs on linux 

https://github.com/zfsonlinux/zfs

Perhaps I could help to research information’s. The most Customers from us uses Proxmox which contain prebuild zfs.  What did you mean with library ? 

 Nice would be if netdata could check once per day  if a pool is deprecated or all disk are online and bring up an alert

 

Zfs pool states listed here: 

https://docs.joyent.com/private-cloud/troubleshooting/disk-replacement

 

So if netdata could check the output for these states and would get a result we would only need a alert

 

zpool status | grep DEGRADED

 

nice would be to see if all pools are online but important would be an alert for the states DEGRADED FAULTED which would need a daily check

 

best Regards

==

 

ONLINE

All devices can (and should while operating optimally) be in the ONLINE state. This includes the pool, top-level VDEVs (parity groups of type mirror, raidz{1,2,3}) and the drives themselves. Transitory errors may still occur without the drive changing state.

OFFLINE

Only bottom-level devices (drives) can be OFFLINE. This is a manual administrative state, and healthy drives can be brought back online and active into the pool.

UNAVAIL

The device (or VDEV) in question can not be opened. If a VDEV is UNAVAIL, the pool will not be accessible or able to be imported. UNAVAIL devices may also report as FAULTED in some scenarios. Operationally, UNAVAIL disks are roughly equivelent to FAULTED disks.

DEGRADED

A fault in a device has occurred, impacting all VDEVs above it. The pool is still operable, but redundancy may have been lost in a VDEV.

REMOVED

The device was physically removed while the system was running. Device removal detection is hardware-dependent and might not be supported on all platforms.

FAULTED

All components (top and redundancy VDEVs, and drives) of the pool can be in a FAULTED state. A FAULTED component is completely inaccessible. The severity of a device being DEGRADED depends a lot on which device it is.

INUSE

 

 

Von: Markos Fountoulakis <notifications@github.com>
Antworten an: netdata/netdata <reply@reply.github.com>
Datum: Dienstag, 7. Mai 2019 um 10:48
An: netdata/netdata <netdata@noreply.github.com>
Cc: thefiredragon <danny.faller@googlemail.com>, Mention <mention@noreply.github.com>
Betreff: Re: [netdata/netdata] zfs pool (harddrive) notifications (#5590)

 

There is no ZFS library for linux yet though.

**Comment by mfundul** [comment]
A ZFS user space development library would by something like the `libzfslinux-dev` package of Ubuntu. We would not be calling `zpool` or `zfs` binaries from netdata, but instead we would link to such a library. The research that needs to be done is how to use such a library and how available it is in various operating systems and linux distributions.

**Comment by E46-S54** [comment]
This would be super helpful, following with interest

**Comment by thefiredragon** [comment]
Yes but I'm not a developer so I can't provide the informations for the
library's... :/

Aaron S <notifications@github.com> schrieb am Fr., 20. Dez. 2019, 13:32:

**Comment by skurfuerst** [comment]
Hey :) 

You could easily simply read `/proc/spl/kstat/zfs/<zfs-pool-name>/state`, which is a string `ONLINE` (all good), `DEGRADED` (e.g. a disk failed; but the storage pool operates normally), ... - full reference at https://docs.oracle.com/cd/E19253-01/819-5461/gamno/index.html

I don't know how to fetch the pools via the proc filesystem; but one could simply check for a "state" file by searching subdirectories in `/proc/spl/kstat/zfs/`.

Anything I can do to help here? :) I can also develop etc, as I need this feature anyways.

How would you map this status into a chart or into alerts in Netdata?

All the best,
Sebastian

**Comment by skurfuerst** [comment]
ah... one would count the number of unhealthy pools... and maybe additionally the number of healthy pools as well? then one has an overview how many pools exist; and how many of them in unhealthy state.

**Comment by ilyam8** [comment]
Hi @skurfuerst 

> I can also develop etc, as I need this feature anyways.

Sounds good 👍 

> How would you map this status into a chart or into alerts in Netdata?

`ONLINE` is 0, `DEGRADED` is 1, etc.

**Comment by skurfuerst** [comment]
@ilyam8 do you have a small pointer where this would be roughly added codewise? A few bullet points / code pointers would be awesome :)

**Comment by ilyam8** [comment]
@vlvkobal please ^^

**Comment by vlvkobal** [comment]
If root access isn't needed to read these files, then [`collectors/proc.plugin`](https://github.com/netdata/netdata/tree/master/collectors/proc.plugin) is a proper place for the code. We have arcstats in [`proc_spl_kstat_zfs.c`](https://github.com/netdata/netdata/blob/master/collectors/proc.plugin/proc_spl_kstat_zfs.c) but it might be better to have a separate file for ZFS pool status.

**Comment by skurfuerst** [comment]
@vlvkobal thanks :) Just tested - Root access is not necessary :) I'll try my best in the upcoming weeks....

**Comment by thefiredragon** [comment]
@skurfuerst  any updates to this here?

**Comment by krunkcraig** [comment]
@skurfuerst thanks for working out this enhancement, is there any progress?

**Comment by skurfuerst** [comment]
sorry, did not manage to do anything yet :( Sorry for that...

**Comment by ThoSap** [comment]
This feature would be awesome! 😍

**Comment by thefiredragon** [comment]
For this feature we would also donate some coins if someone could handle this, best Regards

**Comment by thefiredragon** [comment]
Its allowed to bump here? 
This Feature of Monitoring and alerting would be really helpful also with the New netdata cloud Integration

**Comment by pgassmann** [comment]
The zfs project meanwhile has consolidated its efforts in a common `openzfs` project for all operating systems. 

Recently, they added a feature for monitoring with influxdb. But this does not seem to be released yet. But it still might be helpful to find the right metrics and how to interpret them: 
https://github.com/openzfs/zfs/tree/master/cmd/zpool_influxdb

@skurfuerst do you still have plans to work on this?

**Comment by pgassmann** [comment]
I'm also linking the zfs collector from prometheus node_exporter:

* https://github.com/prometheus/node_exporter/blob/master/collector/zfs.go
* https://github.com/prometheus/node_exporter/blob/master/collector/zfs_linux.go
* https://github.com/prometheus/node_exporter/commit/bb8b3fca88538fde29266965a9936db10ab93405

**Comment by pgassmann** [comment]
@ilyam8 There is no progress on zpool monitoring, it is assigned to @vlvkobal and @skurfuerst wanted to look at it a year ago. Should this issue be unassigned or assigned to someone, so there is better chance for progress?

**Comment by ilyam8** [comment]
Hi @pgassmann @thefiredragon, if only un/re assigne would help us 😄 

I briefly reread issue and see the following:

- both @vlvkobal and @mfundul (😢) suggest to investigate ZFS library and _find a convenient way to get this information_.
- we have [zfs collector](https://github.com/netdata/netdata/blob/master/collectors/proc.plugin/proc_spl_kstat_zfs.c) and it collects ZFS ARC (Adaptive Replacement Cache) data from
    - `/proc/spl/kstat/zfs/arcstats`
- `node_exporter` has [zfs collector](https://github.com/prometheus/node_exporter/blob/master/collector/zfs.go) and it collects ZFS data from (linux only)
    - `/proc/spl/kstat/zfs/abdstats`
    - `/proc/spl/kstat/zfs/arcstats`
    - `/proc/spl/kstat/zfs/dbuf_stats`
    - `/proc/spl/kstat/zfs/dmu_tx`
    - `/proc/spl/kstat/zfs/dnodestats`
    - `/proc/spl/kstat/zfs/fm`
    - `/proc/spl/kstat/zfs/vdev_cache_stats`
    - `/proc/spl/kstat/zfs/vdev_mirror_stats`
    - `/proc/spl/kstat/zfs/xuio_stats`
    - `/proc/spl/kstat/zfs/zfetchstats`
    - `/proc/spl/kstat/zfs/zil`

---

So the question for me is do we need that ZFS lib or can proceed with `/proc/spl/kstat/zfs/*`? I guess the answer is _no, we don't need ZFS lib_, because it works for `node_exporter`.

Am i missing something @vlvkobal ?

**Comment by ilyam8** [comment]
zpool data is:
 - `/proc/spl/kstat/zfs/*/io`
 - `/proc/spl/kstat/zfs/*/objset-*`
 - `/proc/spl/kstat/zfs/*/state`

The issue is 

> alert me when a pool is failed or degraded 

I think `/proc/spl/kstat/zfs/*/state` is what we need

https://github.com/prometheus/node_exporter/blob/7717702c96f374a072913052ffaa67e30fd2d98b/collector/zfs_linux.go#L44

**Comment by ilyam8** [comment]
@pgassmann @thefiredragon 

can you guys show `/proc/spl/kstat/zfs/*` structure? 

> tree -L 5 /proc/spl/kstat/zfs/

**Comment by pgassmann** [comment]
The available information depends on the zfs version. E.g. the zpool state information is only available in newer versions.

## Ubuntu 20.04, zfs 0.8.3

```
zfs --version
zfs-0.8.3-1ubuntu12.7
zfs-kmod-0.8.3-1ubuntu12.6
```


```
/proc/spl/kstat/zfs/
├── abdstats
├── arcstats
├── dbgmsg
├── dbufs
├── dbufstats
├── dmu_tx
├── dnodestats
├── fletcher_4_bench
├── fm
├── import_progress
├── vdev_cache_stats
├── vdev_mirror_stats
├── vdev_raidz_bench
├── xuio_stats
├── zfetchstats
├── zfsbulk
│   ├── dmu_tx_assign
│   ├── io
│   ├── iostats
│   ├── multihost
│   ├── objset-0x1493
│   ├── objset-0x183
│   ├── objset-0x203
│   ├── objset-0x36
│   ├── objset-0x506
│   ├── objset-0x584
│   ├── objset-0x617
│   ├── objset-0x62
│   ├── objset-0xf87
│   ├── reads
│   ├── state
│   └── txgs
├── zfsdata
│   ├── dmu_tx_assign
│   ├── io
│   ├── iostats
│   ├── multihost
│   ├── objset-0x1183
│   ├── objset-0x17b2
│   ├── objset-0x183
│   ├── objset-0x1aab
│   ├── objset-0x1ab4
│   ├── objset-0x1abd
│   ├── objset-0x1ba2
│   ├── objset-0x1cfe
│   ├── objset-0x1edf
│   ├── objset-0x203
│   ├── objset-0x36
│   ├── objset-0x586
│   ├── objset-0x789
│   ├── objset-0x810
│   ├── objset-0x81b
│   ├── objset-0x924
│   ├── objset-0x92d
│   ├── objset-0x936
│   ├── objset-0xa91
│   ├── objset-0xd85
│   ├── objset-0xd8f
│   ├── reads
│   ├── state
│   └── txgs
└── zil

2 directories, 60 files
```
```
# find /proc/spl/kstat/zfs/ -maxdepth 5  -ls
4026533982      0 dr-xr-xr-x  20 root     root            0 Mar 24 16:07 /proc/spl/kstat/zfs/
4026533987      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/fm
4026533997      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zil
4026533995      0 -rw-------   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/dbufs
4026533989      0 -rw-------   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/dbgmsg
4026533993      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/dmu_tx
4026534011      0 dr-xr-xr-x  28 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk
4026534015      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/io
4026534013      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/txgs
4026534012      0 -rw-------   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/reads
4026534017      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/state
4026534018      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/iostats
4026534016      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/multihost
4026534110      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x36
4026534116      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x62
4026534112      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x183
4026534117      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x203
4026534118      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x506
4026534119      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x584
4026534114      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x617
4026534113      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0xf87
4026534014      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/dmu_tx_assign
4026534115      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsbulk/objset-0x1493
4026534002      0 dr-xr-xr-x 872 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata
4026534006      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/io
4026534004      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/txgs
4026534003      0 -rw-------   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/reads
4026534008      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/state
4026534009      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/iostats
4026534007      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/multihost
4026534019      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x36
4026534020      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x183
4026534021      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x203
4026534109      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x586
4026534027      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x789
4026534091      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x810
4026534153      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x81b
4026534360      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x924
4026534361      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x92d
4026534362      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x936
4026534111      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0xa91
4026534102      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0xd85
4026534022      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0xd8f
4026534005      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/dmu_tx_assign
4026534010      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1183
4026534120      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x17b2
4026534352      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1aab
4026534358      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1ab4
4026534359      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1abd
4026534353      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1ba2
4026534788      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1cfe
4026534121      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfsdata/objset-0x1edf
4026533988      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/abdstats
4026533994      0 -rw-r--r--   1 root     root            0 Mar 24 16:07 /proc/spl/kstat/zfs/arcstats
4026533996      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/dbufstats
4026533991      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/dnodestats
4026533990      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/xuio_stats
4026533992      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/zfetchstats
4026534001      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/import_progress
4026533983      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/fletcher_4_bench
4026533998      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/vdev_cache_stats
4026534000      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/vdev_raidz_bench
4026533999      0 -rw-r--r--   1 root     root            0 Apr  2 20:15 /proc/spl/kstat/zfs/vdev_mirror_stats
```



## Ubuntu 18.04, ZFS  0.7.5

```
apt show zfsutils-linux 
Package: zfsutils-linux
Version: 0.7.5-1ubuntu16.11
```

```
tree -L 5 /proc/spl/kstat/zfs/
/proc/spl/kstat/zfs/
├── abdstats
├── arcstats
├── dbgmsg
├── dbufs
├── dmu_tx
├── dnodestats
├── fletcher_4_bench
├── fm
├── vdev_cache_stats
├── vdev_raidz_bench
├── xuio_stats
├── zfetchstats
├── zfsbulk
│   ├── dmu_tx_assign
│   ├── io
│   ├── multihost
│   ├── reads
│   └── txgs
├── zfsdata
│   ├── dmu_tx_assign
│   ├── io
│   ├── multihost
│   ├── reads
│   └── txgs
└── zil

2 directories, 23 files
```

## Ubuntu 20.04 with latest zfs 2.0.4

on my pc ( https://github.com/saveriomiroddi/zfs-installer) :

```
zfs --version
zfs-2.0.4-0york0~20.04
zfs-kmod-2.0.4-0york0~20.04
```

```
tree -L 5 /proc/spl/kstat/zfs/
/proc/spl/kstat/zfs/
├── abdstats
├── arcstats
├── bpool
│   ├── dmu_tx_assign
│   ├── io
│   ├── iostats
│   ├── multihost
│   ├── objset-0x36
│   ├── reads
│   ├── state
│   └── txgs
├── dbgmsg
├── dbufs
├── dbufstats
├── dmu_tx
├── dnodestats
├── fletcher_4_bench
├── fm
├── import_progress
├── rpool
│   ├── dmu_tx_assign
│   ├── io
│   ├── iostats
│   ├── multihost
│   ├── objset-0x110
│   ├── objset-0x183
│   ├── objset-0x36
│   ├── reads
│   ├── state
│   └── txgs
├── vdev_cache_stats
├── vdev_mirror_stats
├── vdev_raidz_bench
├── xuio_stats
├── zfetchstats
├── zil
└── zstd

2 directories, 35 files
```

**Comment by pgassmann** [comment]
node_exporter creates separate "dimensions" for the metric `node_zfs_zpool_state` with with a label for for the possible states of a zpool and the value 0/1:

![grafik](https://user-images.githubusercontent.com/460192/113443742-d895d980-93f2-11eb-9c05-65df44bed4d5.png)

**Comment by ThoSap** [comment]
# CentOS Stream 8.5 with latest OpenZFS 2.0.4

```
$ rpm -q centos-stream-release
centos-stream-release-8.5-1.el8.noarch
$ zfs --version
zfs-2.0.4-1
zfs-kmod-2.0.4-1
```
```
$ tree -L 5 /proc/spl/kstat/zfs
/proc/spl/kstat/zfs
├── abdstats
├── arcstats
├── dbgmsg
├── dbufs
├── dbufstats
├── dmu_tx
├── dnodestats
├── fletcher_4_bench
├── fm
├── import_progress
├── tank
│   ├── dmu_tx_assign
│   ├── io
│   ├── iostats
│   ├── multihost
│   ├── objset-0x36
│   ├── reads
│   ├── state
│   └── txgs
├── vdev_cache_stats
├── vdev_mirror_stats
├── vdev_raidz_bench
├── xuio_stats
├── zfetchstats
├── zil
└── zstd

1 directory, 25 files
```

**Comment by ilyam8** [comment]
> node_exporter creates separate "dimensions" for the metric node_zfs_zpool_state with with a label for for the possible states of a zpool and the value 0/1

That is a good question.

There are 2 ways to collect such metric.

 - a single metric, the state is a value


Pros: only one metric, likely one chart, easy to follow.
Cons: need to know value->state mapping. Works only for current state, if you try to get average value for a period the result is senseless.

 - a metric per state, the value is boolean

Pros: no need in value->state mapping, average value for a period makes sense.
Cons: number of metrics == number of states, unlikely one chart if there are several objects (number of dimes == number of objects * number of states).

---

To be honest i tend to agree that _a metric per state_ is better and that approach is correct. Mainly because _a single metric_ works fails badly if it is not current state.

However, we usually use _a single metric_, because we have in mind netdata charts and real-time data.

**Comment by ilyam8** [comment]
We don't have experience with ZFS, so we heavily rely on your help, guys. @ThoSap @pgassmann @thefiredragon 

I think we can add ZFS pools/etc monitoring in steps and we almost have all info to start - adding ZFS pools state monitoring. @vlvkobal 


```
/proc/spl/kstat/zfs/
├── abdstats
├── arcstats
├── bpool
│   ├── dmu_tx_assign
│   ├── io
│   ├── iostats
│   ├── multihost
│   ├── objset-0x36
│   ├── reads
│   ├── state
```

- so we need to scan `/proc/spl/kstat/zfs/`
- find all ZFS pools (dirs with state, io, etc.)
- read `state`


---

- How do we want to group metrics?

Taking into accound that later we want to add `io` and `objset-*` for pools.

A set of charts per pool, the pool name is family (dashboard subsection name)?

```cmd
ZFS Pools
 - apool
   - state
   - io
 - bpool
   - state
   - io
...
```

Or a chart per metric?

```cmd
ZFS Pools
 - state
   - apool
   - bpool
 - io
   - apool
   - bpool
...
```

**Comment by ilyam8** [comment]
@pgassmann could you show content of `state`, `io` and `object-*` files?

**Comment by thefiredragon** [comment]
@ilyam8 

This one would be better:
```
ZFS Pools
 - apool
   - state
   - io
 - bpool
   - state
   - io
```


Content


```
DansHomeServer# cat state
ONLINE
DansHomeServer# cat io
25 3 0x00 1 80 8699562112 1533875878811418
nread    nwritten reads    writes   wtime    wlentime wupdate  rtime    rlentime rupdate  wcnt     rcnt
14302569070592 4403381355520 50606149 169663891 135617989332951 25286760005768921 1533875876297488 177605131912704 1008401294973938 1533875876345405 0        0
DansHomeServer# cat objset-0x10d
44 1 0x01 7 2160 19400013175 1533892731059476
name                            type data
dataset_name                    7    vstorage/vms/subvol-108-disk-0
writes                          4    518663
nwritten                        4    292154719
reads                           4    775596
nread                           4    4237355844
nunlinks                        4    1969
nunlinked                       4    1969
DansHomeServer#
```


objset-* define virtual disks which like a virutal hard drive managed by zfs over zvol , linux see this under /dev/zvol as physical disk

**Comment by ilyam8** [comment]
ZFS Pools state (yes, only pools state so far) monitoring added in https://github.com/netdata/netdata/pull/11071#event-4669423367

Guys, few questions:
 -  there is [`INUSE` state](https://docs.joyent.com/private-cloud/troubleshooting/disk-replacement). 
 
> This is a status reserved for spares which have been used to replace a faulted drive.

ZFS pool can't be in that state, only drives. Correct?

 - what alarms do we need? Notify a user when a pool in what state(s)?

**Comment by ilyam8** [comment]
[Determining the Health Status of ZFS Storage Pools](https://docs.oracle.com/cd/E19253-01/819-5461/gamno/index.html)

> The health of a pool is determined from the health of all its top-level virtual devices. If all virtual devices are ONLINE, then the pool is also ONLINE. If any one of the virtual devices is DEGRADED or UNAVAIL, then the pool is also DEGRADED. If a top-level virtual device is FAULTED or OFFLINE, then the pool is also FAULTED. A pool in the FAULTED state is completely inaccessible. No data can be recovered until the necessary devices are attached or repaired. A pool in the DEGRADED state continues to run, but you might not achieve the same level of data redundancy or data throughput than if the pool were online.

I think we need `warn` when DEGRADED and `crit` when `FAULTED` or `UNAVAIL`

**Comment by ThoSap** [comment]
Here you have the official OpenZFS documentation instead of the, for OpenZFS outdated, Oracle documentation ;)
https://openzfs.github.io/openzfs-docs/man/8/zpoolprops.8.html?highlight=degraded#DESCRIPTION
https://openzfs.github.io/openzfs-docs/man/8/zpoolconcepts.8.html?highlight=degraded#Device_Failure_and_Recovery
https://openzfs.github.io/openzfs-docs/man/5/zfs-events.5.html?highlight=degraded#PAYLOADS

**Comment by ilyam8** [comment]
Thanks, @ThoSap, this documentation seems better, indeed!

**Comment by ilyam8** [comment]
ZFS pools state monitoring was added in https://github.com/netdata/netdata/pull/11071.

It is in the master branch/nightly release, will be in the upcoming release. Any feedback is appreciated!

If there is an interest we can proceed with ZFS pools - I guess the next step could be collection `io` metrics?

**Comment by ilyam8** [comment]
@ThoSap @thefiredragon  `zfs list` command returns per pool size/alloc/free data

Copying from the https://github.com/netdata/netdata/issues/3089#issuecomment-350450582

```cmd
$ zpool list
NAME    SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
data1   1.81T   107G  1.71T         -     3%     5%  1.00x  ONLINE  -
data2   5.44T  3.55T  1.88T         -    38%    65%  1.00x  ONLINE  -
```

Any ideas where these metrics come from?

My guess is aggregated data from `objset-*` files?

And what are these metrics?

```cmd
DansHomeServer# cat objset-0x10d
44 1 0x01 7 2160 19400013175 1533892731059476
```

**Comment by thefiredragon** [comment]
Opened for this an issue:
https://github.com/openzfs/zfs/discussions/12080

**Comment by frukto** [comment]
> ```batchfile
> $ zpool list
> NAME    SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
> data1   1.81T   107G  1.71T         -     3%     5%  1.00x  ONLINE  -
> data2   5.44T  3.55T  1.88T         -    38%    65%  1.00x  ONLINE  -
> ```
> 
> Any ideas where these metrics come from?
> 
> My guess is aggregated data from `objset-*` files?

I think they are aggregated somewhere arround here https://github.com/openzfs/zfs/blob/master/module/zfs/spa.c#L307

> And what are these metrics?
> 
> ```batchfile
> DansHomeServer# cat objset-0x10d
> 44 1 0x01 7 2160 19400013175 1533892731059476
> ```
The are seem to be somewhat unrelated, other files in `/proc/spl/kstat/zfs/` have similiar headers. And the last value is clearly a time stamp, the last update presumably.

**Comment by thefiredragon** [comment]
More interesting would be if we could provide in netdata more informations about zfs, like Pool size, free Space of the Pool and which drives are in a Pool like the output from zpool status or zpool List

**Comment by putnam** [comment]
We're coming up on a year here -- is there anything stopping netdata from just using zfs list / zpool list? These utilities are designed with scripting in mind. I have the same problem where I get inundated with emails by netdata when inodes/disk consumption look to be trending a certain way--one for every single dataset!

**Comment by ilyam8** [comment]
@cakrit the OP issue is fixed in #11071 @putnam (correct me if I am wrong) is asking about usage/utilization metrics for ZFS pools.

**Comment by ilyam8** [comment]
I am closing this issue because the OP request was implemented a long time ago. If somebody is interested in addition statistics for ZFS pools - please [open a feature request](https://github.com/netdata/netdata/issues/new?assignees=&labels=feature+request%2Cneeds+triage&template=FEAT_REQUEST.yml&title=%5BFeat%5D%3A+).

>  is there anything stopping netdata from just using zfs list / zpool list?

Nothing. I have an access to a Proxmox server with ZFS, so it will be very easy to implement.
